gapis/api/vulkan/framegraph.go

```diff
@@ -16,14 +16,153 @@ package vulkan
+			if rpInfo != nil {
+				panic("Renderpass starts without having ended")
+			}
+			rpInfo = &renderpassInfo{
+				id:       currRpId,
+				beginIdx: subCmdIdx,
+				nodes:    []dependencygraph2.NodeID{},
+				deps:     make(map[uint64]struct{}),
+			}
+			currRpId++
+		}
+
+		// Process commands that are inside a renderpass
+		if rpInfo != nil {
+			nodeID := dependencyGraph.GetCmdNodeID(api.CmdID(subCmdIdx[0]), subCmdIdx[1:])
+			rpInfo.nodes = append(rpInfo.nodes, nodeID)
+		}
+
+		// Ending of renderpass
+		if _, ok := cmdArgs.(VkCmdEndRenderPassArgsʳ); ok {
+			if rpInfo == nil {
+				panic("Renderpass ends without having started")
+			}
+			rpInfo.endIdx = subCmdIdx
+			rpInfos = append(rpInfos, rpInfo)
+			rpInfo = nil
+		}
+	}
+
+	// Iterate on the capture commands to gather information
+	c, err := capture.ResolveGraphicsFromPath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+	if err := sync.MutateWithSubcommands(ctx, p, c.Commands, nil, nil, postSubCmdCb); err != nil {
+		return nil, err
+	}
+
+	updateDependencies(rpInfos, dependencyGraph)
+
+	// Build the framegraph nodes and edges from collected data.
+	nodes := make([]*api.FramegraphNode, len(rpInfos))
+	for i, rpInfo := range rpInfos {
+		// Graphviz DOT: use "\l" as a newline to obtain left-aligned text.
+		text := fmt.Sprintf("Renderpass %v\\lbegin:%v\\lend:%v\\l", rpInfo.id, rpInfo.beginIdx, rpInfo.endIdx)
+		nodes[i] = &api.FramegraphNode{
+			Id:   rpInfo.id,
+			Text: text,
+		}
+	}
+
+	edges := []*api.FramegraphEdge{}
+	for _, rpInfo = range rpInfos {
+		for deps := range rpInfo.deps {
+			edges = append(edges, &api.FramegraphEdge{
+				// We want the graph to show the flow of how the frame is
+				// created (rather than the flow of dependencies), so use the
+				// dependency as the edge origin and rpInfo as the destination.
+				Origin:      deps,
+				Destination: rpInfo.id,
+			})
+		}
+	}
+
+	return &api.Framegraph{Nodes: nodes, Edges: edges}, nil
+}
+
+// updateDependencies establishes dependencies between renderpasses.
+func updateDependencies(rpInfos []*renderpassInfo, dependencyGraph dependencygraph2.DependencyGraph) {
+	// isInsideRenderpass: node -> renderpass it belongs to.
+	isInsideRenderpass := map[dependencygraph2.NodeID]uint64{}
+	for _, rpInfo := range rpInfos {
+		for _, n := range rpInfo.nodes {
+			isInsideRenderpass[n] = rpInfo.id
+		}
+	}
+	// node2renderpasses: node -> set of renderpasses it depends on.
+	node2renderpasses := map[dependencygraph2.NodeID]map[uint64]struct{}{}
+
+	// For a given renderpass RP, for each of its node, explore the dependency
+	// graph in reverse order to mark all the nodes dependending on RP until we
+	// hit the node of another renderpass, which then depends on RP.
+	for _, rpInfo := range rpInfos {
+		// markNode is recursive, so declare it before initializing it.
+		var markNode func(dependencygraph2.NodeID) error
+		markNode = func(node dependencygraph2.NodeID) error {
```

Comment:
I don't know how does go compiler handles this(maybe realises and optimizes) but instead of creating a new lambda function for every rpInfo in rpInfos this might be converted into a normal recursive function calls. But not the highest priority I would say especially if it will take long time.

Reply:
There are O(100) renderpasses per frame on realistic workloads, so I don't think we need to worry there. If framegraph creation turns out to be slow, we'll measure and optimize accordingly.